src/circuit_proof/r1cs.rs

```diff
@@ -53,112 +65,154 @@ impl LinearCombination {
-        self.b.push(lc_b);
-        self.c.push(lc_c);
-        Ok(())
+        aL_val: Assignment,
+        aR_val: Assignment,
+        aO_val: Assignment,
+    ) -> (Variable, Variable, Variable) {
+        let aL_var = self.make_variable(VariableType::aL, aL_val);
+        let aR_var = self.make_variable(VariableType::aR, aR_val);
+        let aO_var = self.make_variable(VariableType::aO, aO_val);
+        (aL_var, aR_var, aO_var)
     }
 
-    fn eval_lc(&self, lc: &LinearCombination) -> Scalar {
-        let sum_vars: Scalar = lc
-            .variables
-            .iter()
-            .map(|(var, scalar)| scalar * self.var_assignment[var.0])
-            .sum();
-        sum_vars + lc.constant
+    // Allocate a variable for v, and assign it the Result value passed in.
+    // Prover will pass in Ok(Scalar), and Verifier will pass in R1CSError.
+    pub fn assign_v(&mut self, v_val: Assignment) -> Variable {
+        self.make_variable(VariableType::v, v_val)
     }
 
-    fn create_verifier_input(
+    fn make_variable(&mut self, var_type: VariableType, value: Assignment) -> Variable {
+        let index = match var_type {
+            VariableType::aL => {
+                self.aL_assignment.push(value);
+                self.aL_assignment.len() - 1
+            }
+            VariableType::aR => {
+                self.aR_assignment.push(value);
+                self.aR_assignment.len() - 1
+            }
+            VariableType::aO => {
+                self.aO_assignment.push(value);
+                self.aO_assignment.len() - 1
+            }
+            VariableType::v => {
+                self.v_assignment.push(value);
+                self.v_assignment.len() - 1
+            }
+        };
+        Variable { var_type, index }
+    }
+
+    // Get number of multiplications.
+    pub fn get_n(&self) -> usize {
+        let n = self.aL_assignment.len();
+        if n == 0 || n.is_power_of_two() {
+            return n;
+        }
+        return n.next_power_of_two();
+    }
+
+    // Get number of high-level witness variables.
+    pub fn get_m(&self) -> usize {
+        self.v_assignment.len()
+    }
+
+    pub fn make_V(
         &self,
-        v_blinding: &Vec<Scalar>,
         pedersen_gens: &PedersenGenerators,
-    ) -> VerifierInput {
-        let V: Vec<RistrettoPoint> = self
-            .var_assignment
+        v_blinding: &Vec<Scalar>,
+    ) -> Result<Vec<RistrettoPoint>, R1CSError> {
+        if v_blinding.len() != self.get_m() {
+            return Err(R1CSError::IncorrectInputSize);
+        }
+        self.v_assignment
             .iter()
             .zip(v_blinding)
-            .map(|(v_i, v_blinding_i)| pedersen_gens.commit(*v_i, *v_blinding_i))
-            .collect();
-        VerifierInput::new(V)
+            .map(|(v_i, v_blinding_i)| Ok(pedersen_gens.commit(v_i.clone()?, *v_blinding_i)))
+            .collect()
+    }
+
+    pub fn constrain(&mut self, lc: LinearCombination) {
```

Comment:
Very subjective nitpick: `add_constraint` would read more clearly to me at the call site.